eth/executionclient/multi_client.go

```diff
@@ -0,0 +1,365 @@
+		}
+		if mc.chainID == nil {
+			mc.chainID = chainID
+			continue
+		}
+		if mc.chainID.Cmp(chainID) != 0 {
+			mc.logger.Error("chain ID mismatch",
+				zap.String("observed_chain_id", mc.chainID.String()),
+				zap.String("checked_chain_id", chainID.String()),
+				zap.String("address", addr))
+			return false, nil
+		}
+	}
+
+	return true, nil
+}
+
+// FetchHistoricalLogs retrieves historical logs emitted by the contract starting from fromBlock.
+// It calls FetchHistoricalLogs of all clients until a no-error result.
+// It doesn't handle errors in the error channel to simplify logic.
+// In this case, caller should call Panic/Fatal to restart the node.
+func (mc *MultiClient) FetchHistoricalLogs(ctx context.Context, fromBlock uint64) (<-chan BlockLogs, <-chan error, error) {
+	var logCh <-chan BlockLogs
+	var errCh <-chan error
+
+	f := func(client SingleClientProvider) (any, error) {
+		singleLogsCh, singleErrCh, err := client.FetchHistoricalLogs(ctx, fromBlock)
+		if err != nil {
+			return nil, err
+		}
+
+		logCh = singleLogsCh
+		errCh = singleErrCh
+		return nil, nil
+	}
+
+	_, err := mc.call(contextWithMethod(ctx, "FetchHistoricalLogs"), f)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return logCh, errCh, nil
+}
+
+// StreamLogs subscribes to events emitted by the contract.
+// NOTE: StreamLogs spawns a goroutine which calls os.Exit(1) if no client is available.
+func (mc *MultiClient) StreamLogs(ctx context.Context, fromBlock uint64) <-chan BlockLogs {
+	logs := make(chan BlockLogs)
+
+	go func() {
+		defer close(logs)
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-mc.closed:
+				return
+			default:
+				// Update healthyCh of all nodes and make sure at least one of them is available.
+				if err := mc.Healthy(ctx); err != nil {
+					mc.logger.Fatal("no healthy clients", zap.Error(err))
+				}
+
+				f := func(client SingleClientProvider) (any, error) {
+					lastBlock, err := client.streamLogsToChan(ctx, logs, fromBlock)
+					if errors.Is(err, ErrClosed) || errors.Is(err, context.Canceled) {
+						return lastBlock, err
+					}
+					if err != nil {
+						// fromBlock's value in the outer scope is updated here, so this function needs to be a closure
+						fromBlock = max(fromBlock, lastBlock+1)
+						return nil, err
+					}
+
+					return nil, nil
+				}
+
+				_, err := mc.call(contextWithMethod(ctx, "StreamLogs"), f)
+				if err != nil && !errors.Is(err, ErrClosed) && !errors.Is(err, context.Canceled) {
+					// NOTE: There are unit tests that trigger Fatal and override its behavior.
+					// Therefore, the code must call `return` afterward.
+					mc.logger.Fatal("failed to stream registry events", zap.Error(err))
+				}
+				return
+			}
+		}
+	}()
+
+	return logs
+}
+
+// Healthy returns if execution client is currently healthy: responds to requests and not in the syncing state.
+func (mc *MultiClient) Healthy(ctx context.Context) error {
+	atLeastOne := atomic.Bool{}
```

Comment:
@moshe-blox @y0sher I think `atLeastOne` is not clear enough, let's rename it to `atLeastOneHealthy`?